swarm/api/filesystem.go

```diff
@@ -110,40 +109,52 @@ func (fs *FileSystem) Upload(lpath, index string, toEncrypt bool) (string, error
 		}
 		awg.Add(1)
 		go func(i int, entry *manifestTrieEntry, done chan bool) {
-			f, err := os.Open(entry.Path)
-			if err == nil {
-				stat, _ := f.Stat()
-				var hash storage.Address
-				var wait func(context.Context) error
-				ctx := context.TODO()
-				hash, wait, err = fs.api.fileStore.Store(ctx, f, stat.Size(), toEncrypt)
-				if hash != nil {
-					list[i].Hash = hash.Hex()
-				}
-				err = wait(ctx)
+			defer func() {
 				awg.Done()
-				if err == nil {
-					first512 := make([]byte, 512)
-					fread, _ := f.ReadAt(first512, 0)
-					if fread > 0 {
-						mimeType := http.DetectContentType(first512[:fread])
-						if filepath.Ext(entry.Path) == ".css" {
-							mimeType = "text/css"
-						}
-						list[i].ContentType = mimeType
-					}
-				}
-				f.Close()
+				done <- true
+			}()
+
+			f, err := os.Open(entry.Path)
+			if err != nil {
+				errors[i] = err
+				return
+			}
+			defer f.Close()
+
+			stat, err := f.Stat()
+			if err != nil {
+				errors[i] = err
+				return
+			}
+
+			var hash storage.Address
+			var wait func(context.Context) error
+			ctx := context.TODO()
+			hash, wait, err = fs.api.fileStore.Store(ctx, f, stat.Size(), toEncrypt)
+			if hash != nil {
+				list[i].Hash = hash.Hex()
+			}
+
+			list[i].ContentType, err = DetectContentType(f)
+			if err != nil {
+				errors[i] = err
+				return
+			}
+
+			if err := wait(ctx); err != nil {
```

Comment:
I'd rather have this `wait()` before the mime detection happens, as in the original code